### PR TITLE
feat(jans-cedarling): add support for the optional SSA JWT

### LINKED ISSUE #10848: feat(jans-cedarling): add support for the optional SSA JWT
-------------------

### Description

- Development target issue

target issue: #10795 

- Implementation Details

When obtaining an `access_token` for sending requests to the lock server's `/audit` endpoint, a Software Statement Assertion (SSA) JWT might be required by the IDP. The SSA JWT must be send in together with the request for DCR.

-------------------

### Prepare

- [x] Read contribution guidelines
- [x] Read license information

------------------

### Identified code changes

- [ ] Pass in the SSA JWT with the request to the DCR if it is available.

-------------------

### Test cases and code coverage

- [ ] Write unit test to cover added/changed code
- [ ] Update integration tests to cover added/changed code

-------------------

### Document the changes

- [ ] task for updating user guides if needed
- [ ] task for updating installation and configuration guides if needed 
- [ ] task for updating developer documentation if needed
- [ ] task for updating technical documentation if needed

### PR DESCRIPTION
**Pull request by 0xtinkle**
### Prepare

- [ ] Read [PR guidelines](https://github.com/JanssenProject/jans/blob/main/docs/CONTRIBUTING.md#prs)
- [ ] Read [license information](https://github.com/JanssenProject/jans/blob/main/LICENSE)

-------------------

#### Target issue

closes #10848 

#### Implementation Details
When obtaining an access_token for sending requests to the lock server's /audit endpoint, a Software Statement Assertion (SSA) JWT might be required by the IDP. The SSA JWT must be send in together with the request for DCR.

-------------------
### Test and Document the changes
- [ ] Static code analysis has been run locally and issues have been fixed
- [ ] Relevant unit and integration tests have been added/updated
- [ ] Relevant documentation has been updated if any (i.e. user guides, installation and configuration guides, technical design docs etc)

Please check the below before submitting your PR. The PR will not be merged if there are no commits that start with `docs:` to indicate documentation changes or if the below checklist is not selected.
- [ ] **I confirm that there is no impact on the docs due to the code changes in this PR.**

### CODE REVIEW

**Comment by mo-auto** [comment]
### :tada: **Snyk checks have passed. No issues have been found so far.**

:white_check_mark: **security/snyk** check is complete. No issues have been found. [(View Details)](https://app.snyk.io/org/janssenproject/pr-checks/c5215151-ed11-4fc1-86e1-cd5bb5f775e9?refs=edc013d0c11296035f58125df07bc8fcc5fc640c&source=prChecksComment)

File: jans-cedarling/cedarling/src/lock/ssa_validation.rs [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+        _ => Err(SsaValidationError::UnsupportedAlgorithm),
+    }
+}
+
+/// Validates the SSA JWT signature and claims
+fn validate_ssa_signature_and_claims(
+    ssa_jwt: &str,
+    decoding_key: &DecodingKey,
+) -> Result<SsaClaims, SsaValidationError> {
+    let mut validation = Validation::new(Algorithm::RS256);
```
Comment (haileyesus2433) [inline jans-cedarling/cedarling/src/lock/ssa_validation.rs]: the signature validation is hardcoded to RS256 can we update the implementation to use the algorithm from the JWT header please

File: jans-cedarling/cedarling/src/lock/ssa_validation.rs [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+                }
+            }
+        }
+    }
+    
+    Err(SsaValidationError::KeyNotFound(kid.clone()))
+}
+
+/// Creates a decoding key from JWK
+fn create_decoding_key(jwk: &Value, algorithm: &Algorithm) -> Result<DecodingKey, SsaValidationError> {
```
Comment (haileyesus2433) [inline jans-cedarling/cedarling/src/lock/ssa_validation.rs:363]: the function doesn't validate that the JWK's alg field matches the JWT header's algorithm and this could lead to algorithm confusion [attacks](https://portswigger.net/web-security/jwt/algorithm-confusion)

File: jans-cedarling/cedarling/src/lock/ssa_validation.rs [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+    let decoded_jwt = decode_jwt(ssa_jwt)?;
+    validate_ssa_structure(&decoded_jwt)?;
+    // Return dummy claims for test
+    let claims: SsaClaims = serde_json::from_value(decoded_jwt.claims.inner.clone())
+        .map_err(|e| SsaValidationError::JwtError(jsonwebtoken::errors::Error::from(e)))?;
+    Ok(claims)
+}
+
+#[cfg(not(test))]
+pub async fn validate_ssa_jwt(
```
Comment (haileyesus2433) [inline jans-cedarling/cedarling/src/lock/ssa_validation.rs]: the function decodes the jwt twice once in decode_jwt and again in validate_ssa_signature_and_claims,
is the double decoding intentional for some reason i might have missed? if not i think we can use the already decoded JWT.

File: jans-cedarling/cedarling/src/lock/ssa_validation.rs [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+    #[error("exp must be a number")]
+    InvalidExpirationTime,
+    
+    #[error("iat must be a number")]
+    InvalidIssuedAtTime,
+    
+    #[error("failed to initialize HTTP client: {0}")]
+    HttpClientError(#[from] reqwest::Error),
+    
+    #[error("failed to fetch JWKS: {0}")]
```
Comment (haileyesus2433) [inline jans-cedarling/cedarling/src/lock/ssa_validation.rs]: JwksFetchError and JwksParseError both use reqwest::Error, but they represent different failure modes this could make debugging harder

File: jans-cedarling/cedarling/src/lock/ssa_validation.rs [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+    InvalidKeyFormat,
+    
+    #[error("unsupported algorithm")]
+    UnsupportedAlgorithm,
+    
+    #[error("JWT validation error: {0}")]
+    JwtError(#[from] jwt::errors::Error),
+}
+
+#[cfg(test)]
```
Comment (haileyesus2433) [inline jans-cedarling/cedarling/src/lock/ssa_validation.rs:492]: the test covers basic scenarios but misses edge case like invalid exp/iat types or RSA keys missing components

File: jans-cedarling/cedarling/src/lock/ssa_validation.rs (hunk 3 above)
Comment (0xtinkle) [reply jans-cedarling/cedarling/src/lock/ssa_validation.rs]: double decoding is intentional.
The first decode only extracts header and claims without signature validation to get alg and kid.
The second decode performs cryptographic signature validation.

File: jans-cedarling/cedarling/src/lock/ssa_validation.rs (hunk 4 above)
Comment (0xtinkle) [reply jans-cedarling/cedarling/src/lock/ssa_validation.rs]: This is because the reqwest::Response::json() method returns a reqwet::Error for both network and JSON parsing errors.
However, it doesn't make debugging harder because the error messages for these variants are different and you can see easily if it's a fetch/network error or a parse/JSON error.

Comment (haileyesus2433) [reply jans-cedarling/cedarling/src/lock/ssa_validation.rs]: ok understood

**Comment by haileyesus2433** [comment]
ok everything looks good

**Review by haileyesus2433 — DISMISSED** [review]

**Review by nynymike — APPROVED** [review]
